File: code_combat/8_Cloudrip_Mountain/468-Grim_Determination/grimDetermination.js
// Your goal is to protect Reynaldo

// Find the paladin with the lowest health.
function lowestHealthPaladin() {
    var lowestHealth = 99999;
    var lowestFriend = null;
    var friends = hero.findFriends();
    for(var f=0; f < friends.length; f++) {
        var friend = friends[f];
        if(friend.type != "paladin") { continue; }
        if(friend.health < lowestHealth && friend.health < friend.maxHealth) {
            lowestHealth = friend.health;
            lowestFriend = friend;
        }
    }
    return lowestFriend;
}

function commandPaladin(paladin) {
    // Heal the paladin with the lowest health using lowestHealthPaladin()
    // You can use paladin.canCast("heal") and command(paladin, "cast", "heal", target)
    // Paladins can also shield: command(paladin, "shield")
    // And don't forget, they can attack, too!
    var lowest = lowestHealthPaladin();
    var enemy = paladin.findNearestEnemy();
    if (paladin.canCast("heal") && lowest)
        hero.command(paladin, "cast", "heal", lowest);
    else if (enemy && paladin.health > paladin.maxHealth * 0.5)
        hero.command(paladin, "attack", enemy);
    else
        hero.command(paladin, "shield");
}

function commandFriends() {
    // Command your friends.
    var friends = hero.findFriends();
    for(var i=0; i < friends.length; i++) {
        var friend = friends[i];
        if(friend.type == "peasant") {
            //commandPeasant(friend);
            commandPeasant(friend);
        } else if(friend.type == "griffin-rider") {
            //commandGriffin(friend);
            commandGriffin(friend);
        } else if(friend.type == "paladin") {
            commandPaladin(friend);
        }
    }
}

function commandGriffin(griffin) {
    var enemy = griffin.findNearestEnemy();
    if (enemy)
        hero.command(griffin, "attack", enemy);
}

function commandPeasant(peasant) {
    var item = peasant.findNearestItem();
    if (item)
        hero.command(peasant, "move", item.pos);
}

function summonGriffins() {
    if (hero.gold >= hero.costOf("griffin-rider"))
        hero.summon("griffin-rider");
}

while(true) {
    commandFriends();
    // Summon griffin riders!
    summonGriffins();
}

File: code_combat/8_Cloudrip_Mountain/468-Grim_Determination/grim_determination.py
def lowestHealthPaladin():
    lowestHealth = 99999
    lowestFriend = None
    friends = hero.findFriends()
    for friend in friends:
        if friend.type != "paladin":
            continue
        if friend.health < lowestHealth and friend.health < friend.maxHealth:
            lowestHealth = friend.health
            lowestFriend = friend
    return lowestFriend


def commandPaladin(paladin):
    lowest = lowestHealthPaladin()
    enemy = paladin.findNearestEnemy()
    if paladin.canCast("heal") and lowest:
        hero.command(paladin, "cast", "heal", lowest)
    elif enemy and paladin.health > paladin.maxHealth * 0.5:
        hero.command(paladin, "attack", enemy)
    else:
        hero.command(paladin, "shield")


def commandFriends():
    friends = hero.findFriends()
    for friend in friends:
        if friend.type == "peasant":
            commandPeasant(friend)
        elif friend.type == "griffin-rider":
            commandGriffin(friend)
        elif friend.type == "paladin":
            commandPaladin(friend)


def commandGriffin(griffin):
    enemy = griffin.findNearestEnemy()
    if enemy:
        hero.command(griffin, "attack", enemy)


def commandPeasant(peasant):
    item = peasant.findNearestItem()
    if item:
        hero.command(peasant, "move", item.pos)


def summonGriffins():
    if hero.gold >= hero.costOf("griffin-rider"):
        hero.summon("griffin-rider")


while True:
    commandFriends()
    summonGriffins()
